tools/python_api/src_cpp/py_connection.cpp

```diff
@@ -174,22 +176,140 @@ std::unordered_map<std::string, std::unique_ptr<Value>> transformPythonParameter
+        return canCastPyLogicalType(
+            *VarListType::getChildType(&from), *VarListType::getChildType(&to));
+    } else if (from.getLogicalTypeID() == LogicalTypeID::MAP &&
+        to.getLogicalTypeID() == LogicalTypeID::MAP) {
+        auto fromKeyType = MapType::getKeyType(&from), fromValueType = MapType::getValueType(&to);
+        auto toKeyType = MapType::getKeyType(&to), toValueType = MapType::getValueType(&to);
+        return
+            (canCastPyLogicalType(*fromKeyType, *toKeyType) ||
+            canCastPyLogicalType(*toKeyType, *fromKeyType)) &&
+            (canCastPyLogicalType(*fromValueType, *toValueType) ||
+            canCastPyLogicalType(*toValueType, *fromValueType));
+    } else {
+        auto castCost = function::BuiltInFunctionsUtils::getCastCost(
+            from.getLogicalTypeID(), to.getLogicalTypeID());
+        return castCost != UNDEFINED_CAST_COST;
+    }
+    return false;
+}
+
+static std::unique_ptr<LogicalType> castPyLogicalType(const LogicalType& from, const LogicalType& to) {
+    // assumes from can cast to to
+    if (from.getLogicalTypeID() == LogicalTypeID::MAP) {
+        auto fromKeyType = MapType::getKeyType(&from), fromValueType = MapType::getValueType(&to);
+        auto toKeyType = MapType::getKeyType(&to), toValueType = MapType::getValueType(&to);
+        auto outputKeyType = canCastPyLogicalType(*fromKeyType, *toKeyType) ? toKeyType : fromKeyType;
+        auto outputValueType = canCastPyLogicalType(*fromValueType, *toValueType) ? toValueType : fromValueType;
+        return LogicalType::MAP(
+            std::make_unique<LogicalType>(*outputKeyType), std::make_unique<LogicalType>(*outputValueType));
+    }
+    return std::make_unique<LogicalType>(to);
+}
+
+static void tryConvertPyLogicalType(LogicalType& from, LogicalType& to) {
+    if (canCastPyLogicalType(from, to)) {
+        from = *castPyLogicalType(from, to);
+    } else if (canCastPyLogicalType(to, from)) {
+        from = *castPyLogicalType(to, from);
+    } else {
+        throw RuntimeException(
+            "Cannot convert Python object to Kuzu value : "
+            + from.toString()
+            + " is incompatible with "
+            + to.toString());
+    }
+}
+
+static std::unique_ptr<LogicalType> pyLogicalType(py::handle val) {
     auto datetime_datetime = importCache->datetime.datetime();
     auto time_delta = importCache->datetime.timedelta();
     auto datetime_date = importCache->datetime.date();
     auto uuid = importCache->uuid.UUID();
-    if (val.is_none()) {
-        return Value::createNullValue();
-    } else if (py::isinstance<py::bool_>(val)) {
-        return Value::createValue<bool>(val.cast<bool>());
+    if (py::isinstance<py::bool_>(val)) {
+        return LogicalType::BOOL();
     } else if (py::isinstance<py::int_>(val)) {
-        return Value::createValue<int64_t>(val.cast<int64_t>());
+        int64_t value = val.cast<int64_t>();
+        if (value > INT32_MAX) {
```

Comment:
I would suggest removing the if/else block, we can blindly bind to int64 for now